ehrapy/api/preprocessing/_data_imputation.py

```diff
@@ -3,51 +3,236 @@
-    impute_conditions = np.logical_or(pd.isnull(adata.X), adata.X == "")
-    if not impute_empty_strings:
-        impute_conditions = pd.isnull(adata.X)
-    adata.X[impute_conditions] = replacement
 
-    # scenario 2: Replace all missing values in a subset of columns with the specified value
-    # TODO
+# ===================== Mean Imputation =========================
+
+
+def _mean(adata: AnnData, var_names: list[str] | None) -> AnnData:
+    imputer = SimpleImputer(strategy="mean")
+    try:
+        # impute a subset of columns
+        if isinstance(var_names, list):
+            column_indices = get_column_indices(adata, var_names)
+            adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+        # impute all columns if None passed
+        else:
+            adata.X = imputer.fit_transform(adata.X)
+    # Imputation using median strategy works with numerical data only
+    except ValueError:
+        raise ImputeStrategyNotAvailableError(
+            "Can only impute numerical data using mean strategy. Try to restrict imputation"
+            "to certain columns using var_names parameter or use a different mode."
+        ) from None
+
+    return adata
+
+
+def _median(adata: AnnData, var_names: list[str] | None) -> AnnData:
+    imputer = SimpleImputer(strategy="median")
+    try:
+        # impute a subset of columns
+        if isinstance(var_names, list):
+            column_indices = get_column_indices(adata, var_names)
+            adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+        # impute all columns if None passed
+        else:
+            adata.X = imputer.fit_transform(adata.X)
+    # Imputation using median strategy works with numerical data only
+    except ValueError:
+        raise ImputeStrategyNotAvailableError(
+            "Can only impute numerical data using median strategy. Try to restrict imputation"
+            "to certain columns using var_names parameter or use a different mode."
+        ) from None
+
+    return adata
+
+
+def _most_frequent(adata: AnnData, var_names: list[str] | None) -> AnnData:
+    imputer = SimpleImputer(strategy="most_frequent")
+    # impute a subset of columns
+    if isinstance(var_names, list):
+        column_indices = get_column_indices(adata, var_names)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
+
+    return adata
+
+
+# ===================== KNN Imputation =========================
+
+
+def _knn(adata: AnnData, var_names: list[str] | None) -> AnnData:
+    # numerical only data needs no encoding since KNN Imputation can be applied directly
+    if np.issubdtype(adata.X.dtype, np.number):
+        _knn_impute(adata, var_names)
+    else:
+        # ordinal encoding is used since non-numerical data can not be imputed using KNN Imputation
```

Comment:
Should this be wrapped with a try except where we can revert the changes if something goes wrong?